File: Go/apps/engine/engine.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"

	"game_of_stones/common"
	"game_of_stones/game"
	"game_of_stones/tree"
)

var logFileName = flag.String("log", "", "log file name")
var logFile *os.File

func main() {
	flag.Parse()
	if *logFileName != "" {
		var err error
		logFile, err = os.Create(*logFileName)
		if err != nil {
			fmt.Printf("Failed to open file %q: %#v\n", *logFileName, err)
		}
		defer logFile.Close()
	}
	reader := bufio.NewReader(os.Stdin)
	theGame := game.NewGame()
	theTree := tree.NewTree(theGame)
loop:
	for {
		line, err := reader.ReadString('\n')
		line = strings.TrimSpace(line)
		if err == io.EOF {
			return
		} else if err != nil {
			panic(err)
		}
		log("got %q\n", line)
		terms := strings.Split(line, " ")
		switch terms[0] {
		case "game-name":
			fmt.Printf("game-name %s\n", game.GameName)
		case "move":
			move, err := game.ParseMove(terms[1])
			if err != nil {
				panic(err)
			}
			theTree.CommitMove(move)
			log("%s", theGame)
		case "undo":
			theTree.Reset()
			log("%s", theGame)
		case "respond":
			millis, err := strconv.ParseInt(terms[1], 10, 64)
			if err != nil {
				panic(err)
			}
			timestamp := time.Now()
			maxDuration := time.Duration(millis) * time.Millisecond
			sims := 0
			for {
				dec, done := theTree.Expand()
				if done || dec != common.NoDecision || time.Since(timestamp) > maxDuration {
					break
				}
				sims++
			}
			move := theTree.BestMove()
			theTree.CommitMove(move)
			decision := theGame.Decision().String()
			fmt.Printf("move %s %s\n", move, decision)
			log("move %s %s %d\n", move, decision, sims)
			log("%s\n", theGame)
		case "stop":
			break loop
		}
	}
}

func log(format string, args ...any) {
	if logFile != nil {
		logFile.WriteString(fmt.Sprintf(format, args...))
	}
}

File: Julia/engine.jl
using Dates

function run_engine(name)
    tree = Tree{Move}(20.0)
    game = Game(name)
    while true
        line = strip(readline())
        if line == ""
            continue
        end
        terms = split(line)
        if terms[1] == "game-name"
            println("game-name $name")
        elseif terms[1] == "move"
            commit_move!(tree, game, terms[2])
        elseif terms[1] == "respond"
            expand!(tree, game)
            millis = parse(Int, terms[2])
            deadline = now() + Millisecond(millis)

            while true
                dec, done = expand!(tree, game)
                if done || dec != no_decision || now() > deadline
                    break
                end
            end

            move = best_move(tree, game)
            println("move $move")
            commit_move!(tree, game, "$move")
        elseif terms[1] == "decision"
            dec = decision(game)
            println("decision $(string(dec))")
        elseif terms[1] == "stop"
            break
        end
    end
end
